File: LeetcodeEasyProblems/_1046. Last Stone Weight/Solution.py
class Solution(object):
    def lastStoneWeight(self, stones):
        """
        :type stones: List[int]
        :rtype: int
        """
        # Create a max-heap by inserting negative values
        max_heap = [-stone for stone in stones]
        heapq.heapify(max_heap)
        
        while len(max_heap) > 1:
            # Extract the two largest elements (invert back to positive)
            first = -heapq.heappop(max_heap)
            second = -heapq.heappop(max_heap)
            
            if first != second:
                # If they are not equal, push the difference back into the heap
                heapq.heappush(max_heap, -(first - second))
        
        # If there's a stone left, return its weight (invert back to positive), else return 0
        return -max_heap[0] if max_heap else 0
        

File: LeetcodeEasyProblems/_1046. Last Stone Weight/Solution.java
import java.util.PriorityQueue;
import java.util.Collections;

public class Solution {
    public int lastStoneWeight(int[] stones) {
        // Create a max-heap using a priority queue
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        
        // Add all stones to the max-heap
        for (int stone : stones) {
            maxHeap.add(stone);
        }
        
        // Process the stones until there is one or no stones left
        while (maxHeap.size() > 1) {
            // Extract the two heaviest stones
            int first = maxHeap.poll();
            int second = maxHeap.poll();
            
            // If the stones have different weights, insert the difference back into the heap
            if (first != second) {
                maxHeap.add(first - second);
            }
        }
        
        // Return the last remaining stone or 0 if no stones are left
        return maxHeap.isEmpty() ? 0 : maxHeap.poll();
    }
}
